app/controllers/invites_controller.rb

```diff
@@ -357,11 +357,33 @@ def ensure_new_registrations_allowed
   end
 
   def ensure_not_logged_in
-    if current_user
-      flash[:error] = I18n.t("login.already_logged_in")
-      render layout: 'no_ember'
-      false
+    return if !current_user
+
+    if invite = Invite.find_by(invite_key: params[:id])
+      added_to_group = false
+
+      if invite.groups.present?
+        invite_by_guardian = Guardian.new(invite.invited_by)
+        new_group_ids = invite.groups.pluck(:id) - current_user.group_users.pluck(:group_id)
+        new_group_ids.each do |id|
+          if group = Group.find_by(id: id)
+            if invite_by_guardian.can_edit_group?(group)
+              group.add(current_user)
+              added_to_group = true
+            end
+          end
+        end
+      end
+
+      if topic = invite.topics.first
+        return redirect_to(topic.url) if guardian.can_see?(topic)
```

Comment:
Does it works if the topic is only visible to a group that the user was just added to in the same invite? I mean, will the `group.add(current_user)` already affect the `guardian` in this line?

Reply:
I was sure that it worked, but no, it did not because the list of secured categories is cached in the `User` object. I updated the PR with the fix, added a test and also moved some of the logic out of `before_action` handler.